x-pack/plugins/dataset_quality/public/state_machines/dataset_quality_controller/src/state_machine.ts

```diff
@@ -0,0 +1,228 @@
+                },
+              },
+            },
+            loaded: {},
+          },
+        },
+      },
+    },
+    {
+      actions: {
+        storeTableOptions: assign((_context, event) => {
+          return 'criteria' in event
+            ? {
+                table: event.criteria,
+              }
+            : {};
+        }),
+        storeFlyoutOptions: assign((context, event) => {
+          return 'dataset' in event
+            ? {
+                flyout: {
+                  ...context.flyout,
+                  dataset: event.dataset as FlyoutDataset,
+                },
+              }
+            : {};
+        }),
+        resetFlyoutOptions: assign((_context, _event) => ({ flyout: undefined })),
+        storeDataStreamStats: assign((_context, event) => {
+          return 'data' in event
+            ? {
+                dataStreamStats: event.data as DataStreamStat[],
+              }
+            : {};
+        }),
+        storeDegradedDocStats: assign((_context, event) => {
+          return 'data' in event
+            ? {
+                degradedDocStats: event.data as DegradedDocsStat[],
+              }
+            : {};
+        }),
+        storeDatasetDetails: assign((context, event) => {
+          return 'data' in event
+            ? {
+                flyout: {
+                  ...context.flyout,
+                  datasetDetails: event.data as DataStreamDetails,
+                },
+              }
+            : {};
+        }),
+      },
+    }
+  );
+
+export interface DatasetQualityControllerStateMachineDependencies {
+  initialContext?: DatasetQualityControllerContext;
+  toasts: IToasts;
+  dataStreamStatsClient: IDataStreamsStatsClient;
+}
+
+export const createDatasetQualityControllerStateMachine = ({
+  initialContext = DEFAULT_CONTEXT,
+  toasts,
+  dataStreamStatsClient,
+}: DatasetQualityControllerStateMachineDependencies) =>
+  createPureDatasetQualityControllerStateMachine(initialContext).withConfig({
+    actions: {
+      notifyFetchDatasetStatsFailed: (_context, event: DoneInvokeEvent<Error>) =>
+        fetchDatasetStatsFailedNotifier(toasts, event.data),
+      notifyFetchDegradedStatsFailed: (_context, event: DoneInvokeEvent<Error>) =>
+        fetchDegradedStatsFailedNotifier(toasts, event.data),
+      notifyFetchDatasetDetailsFailed: (_context, event: DoneInvokeEvent<Error>) =>
+        fetchDatasetDetailsFailedNotifier(toasts, event.data),
+    },
+    services: {
+      loadDataStreamStats: (_context) => dataStreamStatsClient.getDataStreamsStats(),
+      loadDegradedDocs: (_context) => {
+        const defaultTimeRange = getDefaultTimeRange();
+
+        return dataStreamStatsClient.getDataStreamsDegradedStats({
+          start: defaultTimeRange.from,
+          end: defaultTimeRange.to,
+        });
+      },
+      loadDataStreamDetails: (context) => {
+        const { type, name: dataset, namespace } = context.flyout.dataset as FlyoutDataset;
```

Comment:
**nit**: what is the for this assertion here? Should this be a guard instead to check whether `flyout.dataset` exist?

Reply:
do you mean something like
```
    if (!context.flyout.dataset) {
      return Promise.reject('Dataset is not specified');
    }

    const { type, name: dataset, namespace } = context.flyout.dataset;

    return dataStreamStatsClient.getDataStreamDetails({
      dataStream: dataStreamPartsToIndexName({
        type: type as DataStreamType,
        dataset,
        namespace,
      }),
    });
```
?

Reply:
Something like that, but if throws an error you should also handle that in the UI to tell the user what is not working.
I was thinking more about something
```ts
if ('dataset' in context.flyout) {
      const { type, name: dataset, namespace } = context.flyout.dataset;

    return dataStreamStatsClient.getDataStreamDetails({
      dataStream: dataStreamPartsToIndexName({
        type: type as DataStreamType,
        dataset,
        namespace,
      }),
    });
 }
```
which guards you against a type failure and doesn't need to throw an error, although you should consider what is the ideal feedback to give if this scenario occurs.

Reply:
Addressed in [66b7dbb](https://github.com/elastic/kibana/pull/174906/commits/66b7dbbeebe835ec361d8b7896ceb664c514e304).